### PR TITLE
[ML] Explain anomaly score factors

### PR DESCRIPTION
**Pull request by valeriy42**
This PR surfaces new information about the impact of the factors on the initial anomaly score in the anomaly record:

- single bucket impact is determined by the deviation between actual and typical in the current bucket
- multi-bucket impact is determined by the deviation between actual and typical in the past 12 buckets
- anomaly characteristics are statistical properties of the current anomaly compared to the historical observations
- high variance penalty is the reduction of anomaly score in the buckets with large confidence intervals.
- incomplete bucket penalty is the reduction of anomaly score in the buckets with fewer samples than historically expected.

Additionally, we compute lower- and upper-confidence bounds and the typical value for the anomaly records. This improves the explainability of the cases where the model plot is not activated with only a slight overhead in performance (1-2%).

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/AnomalyRecord.java [hunk 1]
```diff
@@ -253,6 +261,7 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeList(influences);
         }
         out.writeOptionalWriteable(geoResults);
+        out.writeOptionalWriteable(anomalyScoreExplanation);
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/AnomalyRecord.java:264]: Extra protection is need in a mixed version cluster as pre 8.6 nodes do not know about `anomalyScoreExplanation`

```
if (in.getVersion().onOrAfter(Version.V_8_6_0)) {
   out.writeOptionalWriteable(anomalyScoreExplanation);
}
```

Similarly for the ctor where the field is read

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/AnomalyScoreExplanation.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    private Integer singleBucketImpact;
+    private Integer multiBucketImpact;
+    private Integer anomalyCharacteristicsImpact;
+    private Double lowerConfidenceBound;
+    private Double typicalValue;
+    private Double upperConfidenceBound;
+    private Boolean highVariancePenalty;
+    private Boolean incompleteBucketPenalty;
+
+    public AnomalyScoreExplanation() {}
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/AnomalyScoreExplanation.java:68]: It looks like this ctor is only used in tests and can be package private 

```suggestion
   AnomalyScoreExplanation() {}
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/ReservedFieldNames.java [hunk 3]
```diff
@@ -82,6 +83,17 @@ public final class ReservedFieldNames {
         AnomalyRecord.INITIAL_RECORD_SCORE.getPreferredName(),
         AnomalyRecord.BUCKET_SPAN.getPreferredName(),
 
+        AnomalyScoreExplanation.ANOMALY_TYPE.getPreferredName(),
+        AnomalyScoreExplanation.ANOMALY_LENGTH.getPreferredName(),
+        AnomalyScoreExplanation.SINGLE_BUCKET_IMPACT.getPreferredName(),
+        AnomalyScoreExplanation.MULTI_BUCKET_IMPACT.getPreferredName(),
+        AnomalyScoreExplanation.ANOMALY_CHARACTERISTICS_IMPACT.getPreferredName(),
+        AnomalyScoreExplanation.LOWER_CONFIDENCE_BOUND.getAllReplacedWith(),
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/ReservedFieldNames.java:91]:
```suggestion
        AnomalyScoreExplanation.LOWER_CONFIDENCE_BOUND.getPreferredName(),
```

Comment (valeriy42) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/ReservedFieldNames.java]: Oh, missed this autosuggest error 🙈  Fixed in a0e3327.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/AnomalyScoreExplanation.java (hunk 2 above)
Comment (valeriy42) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/AnomalyScoreExplanation.java]: Fixed in a0e3327

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/AnomalyRecord.java (hunk 1 above)
Comment (valeriy42) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/results/AnomalyRecord.java]: Good point 👍 Fixed in a0e3327.

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by valeriy42** [comment]
@elasticsearchmachine update branch

**Pull request merged**